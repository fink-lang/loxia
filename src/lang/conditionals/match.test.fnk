{fink2js} = import '../../testing'
{describe, it, only, expect, to_match_snapshot} = import '@fink/jest'


describe 'match', fn:
  it 'compiles', fn:
    expect
      fink2js`
        match shrub:
          # simple value
          123:
            # result
            spam

          # object matching
          {foo: 4, ni}:
            spam + ni

          # array matching
          [1, 2, ni]:
            ni

          # fallback if none of above match
          else:
            shrub
      `
      to_match_snapshot


  it 'compiles without else', fn:
    expect
      fink2js`
        match shrub:
          {foo: 4, ni}: spam + ni
          123: spam
      `
      to_match_snapshot


  it 'compiles nested pattern', fn:
    expect
      fink2js`
        match shrub:
          # deep obj comparison
          {foo: {}}: spam + ni
          {foo: []}: spam + ni
          {foo: 4, ni: {na, nu}}: spam + ni
          {xx: 1, foo: {bar: 'spam'}, shrub: {na: 'nu'}}: ni
          {ni: {len: 1}, na: {len: 1}}: na
          # calculated props
          {(a): {b, c}}: ni
          # spread
          {(foo): {}, ...{bar, spam}}: ni
      `
      to_match_snapshot

    expect
      fink2js`
        match shrub:
          # deep array comparison
          [1, 2, [ni]]: ni
          [1, 2, [], {}]: foo
          [1, [2, 3], 4]: bar
          [1, [2, [3, 4]], [5, 6], 7]: spam
      `
      to_match_snapshot

    expect
      fink2js`
        # spread and ignored elems
        match shrub:
          [1, ,2]: foo
          [1, ...{length: 0}]: foo
          [1, ..., 3]: foo
          [1, ...{length: 0}, 3]: foo
      `
      to_match_snapshot


  it 'compiles nested match', fn:
    expect
      fink2js`
        test = fn:
          match foo:
            bar:
              match spam:
                shrub: ni
              bar
      `
      to_match_snapshot


  it 'compiles patial application', fn:
    expect
      fink2js`
        match {spam: 123}:
          # simple unary
          !?: ni

          # simple binary
          ? > 123: ni

          # simple call
          ?(): ni

          # simple as call arg
          shrub ?: ni


          # nested unary
          {spam: !?}: ni

          # nested binary
          {spam: ? > 123}: ni

          # nested call
          {spam: ?()}: ni

          # nested member call
          {spam: ?.is_foo()}: ni

          # nested as call arg
          {spam: is_foo ?}: ni


          # nested unary
          [!?]: ni

          # nested binary
          [? > 123]: ni

          # nested call
          [?()]: ni

          # nested member call
          [?.is_foo()]: ni

          # nested as call arg
          [is_foo ?]: ni
      `
      to_match_snapshot


  it 'compiles do-expr', fn:
    expect
      fink2js`
        foo = match bar:
          spam:
            ni
            throw foobar
          else:
            shrub
            throw ni
      `
      to_match_snapshot

