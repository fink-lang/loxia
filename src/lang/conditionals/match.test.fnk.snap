// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match compiles 1`] = `
"{
  ˆmatch_2: {
    const ˆvalue_1 = shrub;

    // simple value
    if (ˆvalue_1 === 123) {
      spam;
      break ˆmatch_2;
    }

    // object matching
    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null) {
      const {
        \\"foo\\": ˆp_3,
        \\"ni\\": ˆp_4
      } = ˆvalue_1;
      // object matching
      if (ˆp_3 === 4) // object matching
        if (ˆp_4 === ni) {
          spam + ni;
          break ˆmatch_2;
        }
    }

    // array matching
    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null && typeof ˆvalue_1[Symbol.iterator] === \\"function\\") {
      const [ˆa_5, ˆa_6, ˆa_7] = ˆvalue_1;
      // array matching
      if (ˆa_5 === 1) // array matching
        if (ˆa_6 === 2) // array matching
          if (ˆa_7 === ni) {
            ni;
            break ˆmatch_2;
          }
    }

    // fallback if none of above match
    {
      shrub;
      break ˆmatch_2;
    }
  }
}
Object.assign(module.exports, {});"
`;

exports[`match compiles nested match 1`] = `
"const test = () => {
  const ˆvalue_1 = foo;

  if (ˆvalue_1 === bar) {
    {
      {
        ˆmatch_4: {
          const ˆvalue_3 = spam;

          if (ˆvalue_3 === shrub) {
            ni;
            break ˆmatch_4;
          }
        }
      }
      return bar;
    }
  }
};

Object.assign(module.exports, {
  test
});"
`;

exports[`match compiles nested pattern 1`] = `
"{
  ˆmatch_2: {
    const ˆvalue_1 = shrub;

    // deep obj comparison
    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null) {
      const {
        \\"foo\\": ˆp_3
      } = ˆvalue_1;

      // deep obj comparison
      if (ˆp_3 !== undefined && ˆp_3 !== null) {
        spam + ni;
        break ˆmatch_2;
      }
    }

    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null) {
      const {
        \\"foo\\": ˆp_4
      } = ˆvalue_1;

      if (ˆp_4 !== undefined && ˆp_4 !== null && typeof ˆp_4[Symbol.iterator] === \\"function\\") {
        spam + ni;
        break ˆmatch_2;
      }
    }

    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null) {
      const {
        \\"foo\\": ˆp_5,
        \\"ni\\": ˆp_6
      } = ˆvalue_1;
      if (ˆp_5 === 4) if (ˆp_6 !== undefined && ˆp_6 !== null) {
        const {
          \\"na\\": ˆp_7,
          \\"nu\\": ˆp_8
        } = ˆp_6;
        if (ˆp_7 === na) if (ˆp_8 === nu) {
          spam + ni;
          break ˆmatch_2;
        }
      }
    }

    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null) {
      const {
        \\"xx\\": ˆp_9,
        \\"foo\\": ˆp_10,
        \\"shrub\\": ˆp_11
      } = ˆvalue_1;
      if (ˆp_9 === 1) if (ˆp_10 !== undefined && ˆp_10 !== null) {
        const {
          \\"bar\\": ˆp_12
        } = ˆp_10;
        if (ˆp_12 === \`spam\`) if (ˆp_11 !== undefined && ˆp_11 !== null) {
          const {
            \\"na\\": ˆp_13
          } = ˆp_11;

          if (ˆp_13 === \`nu\`) {
            ni;
            break ˆmatch_2;
          }
        }
      }
    }

    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null) {
      const {
        \\"ni\\": ˆp_14,
        \\"na\\": ˆp_15
      } = ˆvalue_1;

      if (ˆp_14 !== undefined && ˆp_14 !== null) {
        const {
          \\"len\\": ˆp_16
        } = ˆp_14;
        if (ˆp_16 === 1) if (ˆp_15 !== undefined && ˆp_15 !== null) {
          const {
            \\"len\\": ˆp_17
          } = ˆp_15;

          if (ˆp_17 === 1) {
            na;
            break ˆmatch_2;
          }
        }
      }
    }

    // calculated props
    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null) {
      const {
        [a]: ˆp_18
      } = ˆvalue_1;

      // calculated props
      if (ˆp_18 !== undefined && ˆp_18 !== null) {
        const {
          \\"b\\": ˆp_19,
          \\"c\\": ˆp_20
        } = ˆp_18;
        // calculated props
        if (ˆp_19 === b) // calculated props
          if (ˆp_20 === c) {
            ni;
            break ˆmatch_2;
          }
      }
    }

    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null) {
      const {
        [foo]: ˆp_21,
        ...ˆp_22
      } = ˆvalue_1;
      if (ˆp_21 !== undefined && ˆp_21 !== null) if (ˆp_22 !== undefined && ˆp_22 !== null) {
        const {
          \\"bar\\": ˆp_23,
          \\"spam\\": ˆp_24
        } = ˆp_22;
        if (ˆp_23 === bar) if (ˆp_24 === spam) {
          ni;
          break ˆmatch_2;
        }
      }
    }
  }
}
Object.assign(module.exports, {});"
`;

exports[`match compiles nested pattern 2`] = `
"{
  ˆmatch_2: {
    const ˆvalue_1 = shrub;

    // deep array comparison
    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null && typeof ˆvalue_1[Symbol.iterator] === \\"function\\") {
      const [ˆa_3, ˆa_4, ˆa_5] = ˆvalue_1;
      // deep array comparison
      if (ˆa_3 === 1) // deep array comparison
        if (ˆa_4 === 2) // deep array comparison
          if (ˆa_5 !== undefined && ˆa_5 !== null && typeof ˆa_5[Symbol.iterator] === \\"function\\") {
            const [ˆa_6] = ˆa_5;

            // deep array comparison
            if (ˆa_6 === ni) {
              ni;
              break ˆmatch_2;
            }
          }
    }

    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null && typeof ˆvalue_1[Symbol.iterator] === \\"function\\") {
      const [ˆa_7, ˆa_8, ˆa_9, ˆa_10] = ˆvalue_1;
      if (ˆa_7 === 1) if (ˆa_8 === 2) if (ˆa_9 !== undefined && ˆa_9 !== null && typeof ˆa_9[Symbol.iterator] === \\"function\\") if (ˆa_10 !== undefined && ˆa_10 !== null) {
        foo;
        break ˆmatch_2;
      }
    }

    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null && typeof ˆvalue_1[Symbol.iterator] === \\"function\\") {
      const [ˆa_11, ˆa_12, ˆa_13] = ˆvalue_1;
      if (ˆa_11 === 1) if (ˆa_12 !== undefined && ˆa_12 !== null && typeof ˆa_12[Symbol.iterator] === \\"function\\") {
        const [ˆa_14, ˆa_15] = ˆa_12;
        if (ˆa_14 === 2) if (ˆa_15 === 3) if (ˆa_13 === 4) {
          bar;
          break ˆmatch_2;
        }
      }
    }

    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null && typeof ˆvalue_1[Symbol.iterator] === \\"function\\") {
      const [ˆa_16, ˆa_17, ˆa_18, ˆa_19] = ˆvalue_1;
      if (ˆa_16 === 1) if (ˆa_17 !== undefined && ˆa_17 !== null && typeof ˆa_17[Symbol.iterator] === \\"function\\") {
        const [ˆa_20, ˆa_21] = ˆa_17;
        if (ˆa_20 === 2) if (ˆa_21 !== undefined && ˆa_21 !== null && typeof ˆa_21[Symbol.iterator] === \\"function\\") {
          const [ˆa_22, ˆa_23] = ˆa_21;
          if (ˆa_22 === 3) if (ˆa_23 === 4) if (ˆa_18 !== undefined && ˆa_18 !== null && typeof ˆa_18[Symbol.iterator] === \\"function\\") {
            const [ˆa_24, ˆa_25] = ˆa_18;
            if (ˆa_24 === 5) if (ˆa_25 === 6) if (ˆa_19 === 7) {
              spam;
              break ˆmatch_2;
            }
          }
        }
      }
    }

    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null && typeof ˆvalue_1[Symbol.iterator] === \\"function\\") {
      const [ˆa_26, ...ˆa_27] = ˆvalue_1;
      if (ˆa_26 === 1) if (ˆa_27 !== undefined && ˆa_27 !== null) {
        const {
          \\"length\\": ˆp_28
        } = ˆa_27;

        if (ˆp_28 === 2) {
          foo;
          break ˆmatch_2;
        }
      }
    }

    // TODO: should just be , ...,
    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null && typeof ˆvalue_1[Symbol.iterator] === \\"function\\") {
      let _do_result;

      {
        const [...ˆitems_32] = ˆvalue_1;
        _do_result = [ˆitems_32, ˆitems_32.slice(1, -1), ˆitems_32.slice(-1)];
      }
      const [[ˆa_29], ˆa_30, [ˆa_31]] = _do_result;
      _do_result = undefined;
      // TODO: should just be , ...,
      if (ˆa_29 === 1) // TODO: should just be , ...,
        if (ˆa_30 !== undefined && ˆa_30 !== null && typeof ˆa_30[Symbol.iterator] === \\"function\\") // TODO: should just be , ...,
          if (ˆa_31 === 3) {
            foo;
            break ˆmatch_2;
          }
    }
  }
}
Object.assign(module.exports, {});"
`;

exports[`match compiles patial application 1`] = `
"{
  ˆmatch_2: {
    const ˆvalue_1 = {
      \\"spam\\": 123
    };

    // simple unary
    if (!ˆvalue_1) {
      ni;
      break ˆmatch_2;
    }

    // simple binary
    if (ˆvalue_1 > 123) {
      ni;
      break ˆmatch_2;
    }

    // simple call
    if (ˆvalue_1()) {
      ni;
      break ˆmatch_2;
    }

    // simple as call arg
    if (shrub(ˆvalue_1)) {
      ni;
      break ˆmatch_2;
    }

    // nested unary
    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null) {
      const {
        \\"spam\\": ˆp_3
      } = ˆvalue_1;

      // nested unary
      if (!ˆp_3) {
        ni;
        break ˆmatch_2;
      }
    }

    // nested binary
    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null) {
      const {
        \\"spam\\": ˆp_4
      } = ˆvalue_1;

      // nested binary
      if (ˆp_4 > 123) {
        ni;
        break ˆmatch_2;
      }
    }

    // nested call
    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null) {
      const {
        \\"spam\\": ˆp_5
      } = ˆvalue_1;

      // nested call
      if (ˆp_5()) {
        ni;
        break ˆmatch_2;
      }
    }

    // nested member call
    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null) {
      const {
        \\"spam\\": ˆp_6
      } = ˆvalue_1;

      // nested member call
      if (ˆp_6.is_foo()) {
        ni;
        break ˆmatch_2;
      }
    }

    // nested as call arg
    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null) {
      const {
        \\"spam\\": ˆp_7
      } = ˆvalue_1;

      // nested as call arg
      if (is_foo(ˆp_7)) {
        ni;
        break ˆmatch_2;
      }
    }

    // nested unary
    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null && typeof ˆvalue_1[Symbol.iterator] === \\"function\\") {
      const [ˆa_8] = ˆvalue_1;

      // nested unary
      if (!ˆa_8) {
        ni;
        break ˆmatch_2;
      }
    }

    // nested binary
    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null && typeof ˆvalue_1[Symbol.iterator] === \\"function\\") {
      const [ˆa_9] = ˆvalue_1;

      // nested binary
      if (ˆa_9 > 123) {
        ni;
        break ˆmatch_2;
      }
    }

    // nested call
    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null && typeof ˆvalue_1[Symbol.iterator] === \\"function\\") {
      const [ˆa_10] = ˆvalue_1;

      // nested call
      if (ˆa_10()) {
        ni;
        break ˆmatch_2;
      }
    }

    // nested member call
    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null && typeof ˆvalue_1[Symbol.iterator] === \\"function\\") {
      const [ˆa_11] = ˆvalue_1;

      // nested member call
      if (ˆa_11.is_foo()) {
        ni;
        break ˆmatch_2;
      }
    }

    // nested as call arg
    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null && typeof ˆvalue_1[Symbol.iterator] === \\"function\\") {
      const [ˆa_12] = ˆvalue_1;

      // nested as call arg
      if (is_foo(ˆa_12)) {
        ni;
        break ˆmatch_2;
      }
    }
  }
}
Object.assign(module.exports, {});"
`;

exports[`match compiles without else 1`] = `
"{
  ˆmatch_2: {
    const ˆvalue_1 = shrub;

    if (ˆvalue_1 !== undefined && ˆvalue_1 !== null) {
      const {
        \\"foo\\": ˆp_3,
        \\"ni\\": ˆp_4
      } = ˆvalue_1;
      if (ˆp_3 === 4) if (ˆp_4 === ni) {
        spam + ni;
        break ˆmatch_2;
      }
    }

    if (ˆvalue_1 === 123) {
      spam;
      break ˆmatch_2;
    }
  }
}
Object.assign(module.exports, {});"
`;
